### ISSUE TITLE
Unify builder and node validation

### ISSUE THREAD

**Issue by freesig** [essential-contributions/essential-node#123]
It would better to reuse the validation code between the validation stream, dry runs and builder. Currently they have different priorities around how state is read. We should investigate how we can unify this or reuse some of the code at least.

**Comment by mitchmindtree** [comment]
Just adding some deets here that I started writing in another issue at the same time:

Currently both the node and builder use `essential_check`, however there are still a few things re-implemented uniquely in both crates:

- Checking a block's `Solution`s in parallel.
- Reading `Predicate`s from state for a `Solution` in parallel.
- `View`s into pre and post state for a given solution index for `essential-check`.

The only difference between the node and builder for these implementations is the approach used to providing access to state (e.g. in the `builder` this is via a combination of `Mutations` and a `ConnectionPool`, in the node's validation it just uses a `ConnectionPool`, in the dry run it uses a combo of two `ConnectionPool`s).

With a single trait along the lines of:

```rust
/// Query the state of a single key at a particular location within a block.
pub trait QueryStateExcl {
    /// The error returned from `query_state_excl`.
    type Error: std::fmt::Debug + std::fmt::Display;
    /// Query the state prior to the solution at the given index within the given block.
    async fn query_state_excl(
        &self,
        contract_ca: &ContentAddress,
        key: &Key,
        block_ca: &ContentAddress,
        solution_ix: SolutionIx,
    ) -> Result<Option<Value>, Self::Error>;
}
```

we should be able to abstract over the source of state, at least for the three cases mentioned above. The top-level fn might look something like:

```rust
pub async fn validate_solutions(
    query_state: &impl QueryStateExcl,
    block_num: BlockNum,
    solutions: impl IntoIterator<Item = (ContentAddress, Arc<Solution>)>,
    conf: &ValidateConfig, // config around parallelism, includes check config, etc
) -> Result<SolutionsSummary, ValidateSolutionsError> {
```

This does assume we have all solutions in memory... We could potentially change `solutions` to also take some minimal trait that asynchronously provides the solutions (similar to a stream or something).
